File: game_ai/Assets/Scripts/AI/PromptBuilder.cs
using System.Text;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

public static class PromptBuilder
{
    public static string Generate(GameStateSnapshot state, List<HandHistory> memory)
    {
        StringBuilder prompt = new StringBuilder();

        // 1. Vai trò và bối cảnh
        prompt.AppendLine("You are a professional Poker AI player in a Texas Hold'em game.");
        prompt.AppendLine($"Current Game Phase: {state.phase}");
        prompt.AppendLine($"Pot Size: {state.potSize} chips. Current Highest Bet to call: {state.highestBet} chips.");

        // 2. Thông tin bài của AI
        string myCards = string.Join(", ", state.me.hand.Select(c => c.displayName));
        prompt.AppendLine($"--- YOUR INFO ---");
        prompt.AppendLine($"Your Name: {state.me.name}");
        prompt.AppendLine($"Your Cards: [{myCards}]");
        prompt.AppendLine($"Your Stack: {state.me.stack} chips. Your Bet this round: {state.me.currentBet}.");

        // 3. Thông tin bài chung
        if (state.communityCards.Count > 0)
        {
            string community = string.Join(", ", state.communityCards.Select(c => c.displayName));
            prompt.AppendLine($"Community Cards on table: [{community}]");
        }

        // 4. Thông tin đối thủ
        prompt.AppendLine("--- OPPONENTS ---");
        foreach (var opp in state.opponents)
        {
            string status = opp.isFolded ? "Folded" : (opp.isAllIn ? "All-In" : "Active");
            prompt.AppendLine($"- {opp.name}: Stack {opp.stack}, Bet {opp.currentBet}, Status: {status}");
        }

        // 5. Diễn biến vòng cược hiện tại (Giúp AI nhận diện độ hung hãn của đối thủ)
        prompt.AppendLine("--- ROUND HISTORY ---");
        if (state.currentRoundHistory.Count > 0)
        {
            foreach (var record in state.currentRoundHistory)
                prompt.AppendLine($"{record.playerName} performed {record.action} ({record.amount} chips)");
        }
        else prompt.AppendLine("No actions yet this round.");

        // 6. TRÍ NHỚ (Học từ các ván trước)
        prompt.AppendLine("--- MEMORY (Past Rounds) ---");
        if (memory.Count > 0)
        {
            // Chỉ gửi 3 ván gần nhất để tránh quá tải Token
            var recentMemory = memory.Skip(Mathf.Max(0, memory.Count - 3)).ToList();
            foreach (var h in recentMemory)
                prompt.AppendLine($"- Round {h.roundNumber}: {h.summary}. Winner: {h.winnerName}");
        }
        else prompt.AppendLine("First round. No memory yet.");

        // 7. Yêu cầu đầu ra (Output format)
        prompt.AppendLine("--- TASK ---");
        prompt.AppendLine("Analyze the board, your hand strength, and opponents' behavior.");
        prompt.AppendLine("Respond ONLY in JSON format with: 'action' (Fold, Check, Call, Raise), 'amount' (if raising), and 'reason' (briefly explain why).");

        return prompt.ToString();
    }
}

File: agent/script/poker_agent/prompt_builder.py
from poker_agent.models import DecisionRequest


SYSTEM_PROMPT = """You are a disciplined Texas Hold'em poker assistant.
You must choose exactly one legal action for the current player.
Use only the legal actions and raise range provided by the input.
Be conservative when the situation is ambiguous.
If action is not Raise, amount must be 0.
Respond with JSON only using keys: action, amount, reason.
"""


def build_user_prompt(request: DecisionRequest) -> str:
    board = ", ".join(card.displayName for card in request.communityCards) or "None"
    my_cards = ", ".join(card.displayName for card in request.me.hand) or "Unknown"

    parts = [
        f"Phase: {request.phase}",
        f"Pot size: {request.potSize}",
        f"Highest bet: {request.highestBet}",
        f"Call amount: {request.callAmount}",
        f"Can fold: {request.canFold}",
        f"Can check: {request.canCheck}",
        f"Can call: {request.canCall}",
        f"Can raise: {request.canRaise}",
        f"Raise amount range: {request.minRaiseAmount} to {request.maxRaiseAmount}",
    ]

    if request.botProfile:
        parts.extend(
            [
                "",
                "[BOT PROFILE]",
                f"Name: {request.botProfile.name}",
                f"Style notes: {request.botProfile.styleNotes or 'None'}",
                f"History window: {request.botProfile.historyWindow}",
                f"Use opponent stats: {request.botProfile.useOpponentStats}",
            ]
        )

    parts.extend(
        [
            "",
            "[ME]",
            f"Name: {request.me.name}",
            f"Stack: {request.me.stack}",
            f"Current bet: {request.me.currentBet}",
            f"Cards: {my_cards}",
            "",
            "[BOARD]",
            board,
            "",
            "[OPPONENTS]",
        ]
    )

    if request.opponents:
        for opponent in request.opponents:
            parts.append(
                f"{opponent.name} | stack={opponent.stack} | bet={opponent.currentBet} | "
                f"folded={opponent.isFolded} | all_in={opponent.isAllIn}"
            )
    else:
        parts.append("None")

    parts.extend(["", "[OPPONENT STATS]"])
    if request.opponentStats:
        for stats in request.opponentStats:
            parts.append(
                f"{stats.name} | hands={stats.handsObserved} | fold={stats.foldRate:.2f} | "
                f"call={stats.callRate:.2f} | raise={stats.raiseRate:.2f} | "
                f"showdown_win={stats.showdownWinRate:.2f} | aggression={stats.aggressionScore:.2f}"
            )
    else:
        parts.append("None")

    parts.extend(["", "[ROUND HISTORY]"])
    if request.currentRoundHistory:
        for action in request.currentRoundHistory:
            parts.append(f"{action.phase}: {action.playerName} -> {action.action} ({action.amount})")
    else:
        parts.append("No actions yet.")

    parts.extend(["", "[RECENT MEMORY]"])
    if request.memory:
        for memory_item in request.memory:
            parts.append(
                f"Round {memory_item.roundNumber}: winner={memory_item.winnerName}, "
                f"pot={memory_item.finalPot}, summary={memory_item.summary}"
            )
    else:
        parts.append("No memory.")

    parts.extend(
        [
            "",
            "[TASK]",
            "Pick the best legal action for this spot.",
            "Return the decision in the structured response format.",
        ]
    )

    return "\n".join(parts)
